File: csharp/TempMail/HttpClient.cs
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using TempMail.Exceptions;

namespace TempMail;

public sealed class HttpClient
{
    private static readonly string[] UaPool = [
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:131.0) Gecko/20100101 Firefox/131.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:130.0) Gecko/20100101 Firefox/130.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:129.0) Gecko/20100101 Firefox/129.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:128.0) Gecko/20100101 Firefox/128.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:127.0) Gecko/20100101 Firefox/127.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14.7; rv:131.0) Gecko/20100101 Firefox/131.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14.6; rv:130.0) Gecko/20100101 Firefox/130.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14.5; rv:129.0) Gecko/20100101 Firefox/129.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14.4; rv:128.0) Gecko/20100101 Firefox/128.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14.3; rv:127.0) Gecko/20100101 Firefox/127.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36 Edg/130.0.0.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36 Edg/129.0.0.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 Edg/127.0.0.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Safari/605.1.15",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.6 Safari/605.1.15",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_4) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4 Safari/605.1.15",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_3) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.3 Safari/605.1.15",
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36",
        "Mozilla/5.0 (X11; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0",
        "Mozilla/5.0 (X11; Linux x86_64; rv:130.0) Gecko/20100101 Firefox/130.0",
        "Mozilla/5.0 (X11; Linux x86_64; rv:129.0) Gecko/20100101 Firefox/129.0",
        "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0",
        "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:130.0) Gecko/20100101 Firefox/130.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36 OPR/116.0.0.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36 OPR/115.0.0.0",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 OPR/114.0.0.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36 OPR/116.0.0.0",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36 OPR/115.0.0.0",
        "Mozilla/5.0 (iPad; CPU OS 17_7 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.6 Mobile/15E148 Safari/604.1",
        "Mozilla/5.0 (iPad; CPU OS 17_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Mobile/15E148 Safari/604.1",
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_7 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.6 Mobile/15E148 Safari/604.1",
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Mobile/15E148 Safari/604.1",
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4 Mobile/15E148 Safari/604.1",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_2) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.2 Safari/605.1.15",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:126.0) Gecko/20100101 Firefox/126.0",
    ];

    private static readonly int[] RetryDelays = [1, 3, 5];
    private const int MaxAttempts = 4;

    private readonly bool _randomUA;
    private readonly CookieContainer _cookieContainer;
    private readonly Random _random = new();

    public System.Net.Http.HttpClient Client { get; }

    public HttpClient(List<string>? proxies = null, bool randomUA = true, bool useCookies = true)
    {
        _randomUA = randomUA;
        _cookieContainer = new CookieContainer();

        var handler = new HttpClientHandler { CookieContainer = _cookieContainer };

        if (proxies is { Count: > 0 })
        {
            handler.UseProxy = true;
            handler.Proxy = new WebProxy(proxies[0]);
        }

        Client = new System.Net.Http.HttpClient(handler);
        Client.Timeout = TimeSpan.FromMinutes(2);
    }

    public CookieContainer Cookies => _cookieContainer;

    public void Dispose() => Client.Dispose();

    // JSON GET with retry
    public async Task<T?> GetFromJsonAsync<T>(string url, JsonSerializerOptions? options = null, CancellationToken ct = default)
    {
        return await RetryAsync(async token =>
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, url);
            SetUA(req);
            using var res = await Client.SendAsync(req, token);
            Check429(res, url);
            res.EnsureSuccessStatusCode();
            return await res.Content.ReadFromJsonAsync<T>(options, token);
        }, ct);
    }

    // JSON POST returning HttpResponseMessage
    public async Task<HttpResponseMessage> PostAsJsonAsync<TValue>(string url, TValue value, JsonSerializerOptions? options = null, CancellationToken ct = default)
    {
        HttpResponseMessage? response = null;
        for (int attempt = 0; attempt < MaxAttempts; attempt++)
        {
            try
            {
                response?.Dispose();
                using var req = new HttpRequestMessage(HttpMethod.Post, url)
                {
                    Content = JsonContent.Create(value, options: options)
                };
                SetUA(req);
                response = await Client.SendAsync(req, ct);
                Check429(response, url);
                response.EnsureSuccessStatusCode();
                return response;
            }
            catch (RateLimitException) when (attempt < RetryDelays.Length)
            {
                response?.Dispose();
                response = null;
                await Task.Delay(TimeSpan.FromSeconds(RetryDelays[attempt]), ct);
            }
            catch
            {
                response?.Dispose();
                throw;
            }
        }
        throw new RateLimitException($"Max retries exceeded for {url}");
    }

    // String GET
    public async Task<string> GetAsync(string url, CancellationToken ct = default)
    {
        return await RetryAsync(async token =>
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, url);
            SetUA(req);
            using var res = await Client.SendAsync(req, token);
            Check429(res, url);
            res.EnsureSuccessStatusCode();
            return await res.Content.ReadAsStringAsync(token);
        }, ct);
    }

    // String GET with header configuration
    public async Task<string> GetAsync(string url, Action<HttpRequestHeaders> configureHeaders, CancellationToken ct = default)
    {
        return await RetryAsync(async token =>
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, url);
            SetUA(req);
            configureHeaders(req.Headers);
            using var res = await Client.SendAsync(req, token);
            Check429(res, url);
            res.EnsureSuccessStatusCode();
            return await res.Content.ReadAsStringAsync(token);
        }, ct);
    }

    // String POST with body string
    public async Task<string> PostAsync(string url, string body, CancellationToken ct = default)
    {
        using var content = new StringContent(body, Encoding.UTF8, "application/json");
        return await RetryAsync(async token =>
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, url) { Content = content };
            SetUA(req);
            using var res = await Client.SendAsync(req, token);
            Check429(res, url);
            res.EnsureSuccessStatusCode();
            return await res.Content.ReadAsStringAsync(token);
        }, ct);
    }

    // String POST with HttpContent
    public async Task<string> PostAsync(string url, HttpContent content, CancellationToken ct = default)
    {
        return await RetryAsync(async token =>
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, url) { Content = content };
            SetUA(req);
            using var res = await Client.SendAsync(req, token);
            Check429(res, url);
            res.EnsureSuccessStatusCode();
            return await res.Content.ReadAsStringAsync(token);
        }, ct);
    }

    // String POST with HttpContent and header configuration
    public async Task<string> PostAsync(string url, HttpContent content, Action<HttpRequestHeaders> configureHeaders, CancellationToken ct = default)
    {
        return await RetryAsync(async token =>
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, url) { Content = content };
            SetUA(req);
            configureHeaders(req.Headers);
            using var res = await Client.SendAsync(req, token);
            Check429(res, url);
            res.EnsureSuccessStatusCode();
            return await res.Content.ReadAsStringAsync(token);
        }, ct);
    }

    // DELETE with retry
    public async Task<HttpResponseMessage> DeleteAsync(string url, CancellationToken ct = default)
    {
        HttpResponseMessage? response = null;
        for (int attempt = 0; attempt < MaxAttempts; attempt++)
        {
            try
            {
                response?.Dispose();
                using var req = new HttpRequestMessage(HttpMethod.Delete, url);
                SetUA(req);
                response = await Client.SendAsync(req, ct);
                Check429(response, url);
                response.EnsureSuccessStatusCode();
                return response;
            }
            catch (RateLimitException) when (attempt < RetryDelays.Length)
            {
                response?.Dispose();
                response = null;
                await Task.Delay(TimeSpan.FromSeconds(RetryDelays[attempt]), ct);
            }
            catch
            {
                response?.Dispose();
                throw;
            }
        }
        throw new RateLimitException($"Max retries exceeded for {url}");
    }

    // Expose DefaultRequestHeaders for provider-level persistent headers
    public HttpRequestHeaders DefaultRequestHeaders => Client.DefaultRequestHeaders;

    private async Task<T> RetryAsync<T>(Func<CancellationToken, Task<T>> fn, CancellationToken ct)
    {
        for (int attempt = 0; attempt < MaxAttempts; attempt++)
        {
            try
            {
                return await fn(ct);
            }
            catch (RateLimitException) when (attempt < RetryDelays.Length)
            {
                await Task.Delay(TimeSpan.FromSeconds(RetryDelays[attempt]), ct);
            }
        }
        throw new RateLimitException("Max retries exceeded");
    }

    private static void Check429(HttpResponseMessage res, string url)
    {
        if ((int)res.StatusCode == 429) throw new RateLimitException($"HTTP 429 from {url}");
    }

    private void SetUA(HttpRequestMessage req)
    {
        if (!_randomUA) return;
        req.Headers.UserAgent.ParseAdd(UaPool[_random.Next(UaPool.Length)]);
    }
}

File: go/providers/httpclient.go
package providers

import (
	"bytes"
	"crypto/tls"
	"io"
	"math/rand"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"strings"
	"sync"
	"time"
)

// 53 user-agent strings to rotate through.
var uaPool = []string{
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:131.0) Gecko/20100101 Firefox/131.0",
	"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:132.0) Gecko/20100101 Firefox/132.0",
	"Mozilla/5.0 (X11; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:130.0) Gecko/20100101 Firefox/130.0",
	"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36 Edg/130.0.0.0",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36 Edg/129.0.0.0",
	"Mozilla/5.0 (iPad; CPU OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1",
	"Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:133.0) Gecko/20100101 Firefox/133.0",
	"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:131.0) Gecko/20100101 Firefox/131.0",
	"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; ARM64; rv:131.0) Gecko/20100101 Firefox/131.0",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 14_0) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
	"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36 Edg/128.0.0.0",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:129.0) Gecko/20100101 Firefox/129.0",
	"Mozilla/5.0 (Android 14; Mobile; rv:131.0) Gecko/131.0 Firefox/131.0",
	"Mozilla/5.0 (iPad; CPU OS 16_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1",
	"Mozilla/5.0 (Linux; Android 14; Pixel 7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Mobile Safari/537.36",
	"Mozilla/5.0 (Linux; Android 14; SM-S918B) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Mobile Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
	"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:128.0) Gecko/20100101 Firefox/128.0",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:131.0) Gecko/20100101 Firefox/131.0",
	"Mozilla/5.0 (Windows NT 11.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 13_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15",
	"Mozilla/5.0 (X11; CrOS x86_64 14541.0.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Linux; Android 13; SM-A546B) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Mobile Safari/537.36",
	"Mozilla/5.0 (Linux; Android 14; moto g73 5G) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Mobile Safari/537.36",
	"Mozilla/5.0 (iPhone; CPU iPhone OS 16_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36 OPR/116.0.0.0",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36 OPR/116.0.0.0",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36 OPR/117.0.0.0",
	"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 Safari/537.36",
	"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Safari/605.1.15",
	"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36",
}

var defaultBackoff = []time.Duration{1 * time.Second, 3 * time.Second, 5 * time.Second}

// HttpClient wraps http.Client with anti-429 features:
// UA rotation, proxy rotation, and automatic retry with backoff.
type HttpClient struct {
	transport *http.Transport
	jar       http.CookieJar
	timeout   time.Duration
	proxies   []string
	randomUA  bool
	mu        sync.Mutex
	rng       *rand.Rand
}

// NewHttpClient creates an HTTP client with optional proxy rotation, random UA, and cookie support.
func NewHttpClient(proxies []string, randomUA bool, useCookies bool) *HttpClient {
	var jar http.CookieJar
	if useCookies {
		jar, _ = cookiejar.New(nil)
	}

	return &HttpClient{
		transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: false},
		},
		jar:      jar,
		timeout:  15 * time.Second,
		proxies:  proxies,
		randomUA: randomUA,
		rng:      rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

// ParseProxies parses a comma-separated string of proxy URLs.
func ParseProxies(s string) []string {
	if s == "" {
		return nil
	}
	parts := strings.Split(s, ",")
	result := make([]string, 0, len(parts))
	for _, p := range parts {
		trimmed := strings.TrimSpace(p)
		if trimmed != "" {
			result = append(result, trimmed)
		}
	}
	return result
}

// Get performs a GET request with retry and proxy rotation.
func (h *HttpClient) Get(url string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, &TempMailError{Provider: "httpclient", Message: err.Error()}
	}
	return h.do(req, nil)
}

// GetWithHeaders performs a GET request with custom headers.
func (h *HttpClient) GetWithHeaders(url string, headers map[string]string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, &TempMailError{Provider: "httpclient", Message: err.Error()}
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	return h.do(req, nil)
}

// Post performs a POST request with retry and proxy rotation.
// Reads body into memory for safe retry.
func (h *HttpClient) Post(url string, contentType string, body io.Reader) ([]byte, error) {
	var bodyBytes []byte
	if body != nil {
		var err error
		bodyBytes, err = io.ReadAll(body)
		if err != nil {
			return nil, &TempMailError{Provider: "httpclient", Message: err.Error()}
		}
	}

	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, &TempMailError{Provider: "httpclient", Message: err.Error()}
	}
	req.Header.Set("Content-Type", contentType)
	return h.do(req, bodyBytes)
}

// PostWithHeaders performs a POST request with custom headers.
func (h *HttpClient) PostWithHeaders(url string, contentType string, body io.Reader, headers map[string]string) ([]byte, error) {
	var bodyBytes []byte
	if body != nil {
		var err error
		bodyBytes, err = io.ReadAll(body)
		if err != nil {
			return nil, &TempMailError{Provider: "httpclient", Message: err.Error()}
		}
	}

	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, &TempMailError{Provider: "httpclient", Message: err.Error()}
	}
	req.Header.Set("Content-Type", contentType)
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	return h.do(req, bodyBytes)
}

// SetCookie adds a cookie to the jar for the given URL.
func (h *HttpClient) SetCookie(rawURL string, cookie *http.Cookie) {
	u, err := url.Parse(rawURL)
	if err != nil || h.jar == nil {
		return
	}
	cookies := h.jar.Cookies(u)
	for i, c := range cookies {
		if c.Name == cookie.Name {
			cookies[i] = cookie
			h.jar.SetCookies(u, cookies)
			return
		}
	}
	h.jar.SetCookies(u, append(cookies, cookie))
}

// GetClient returns a raw *http.Client that shares the same transport and cookie jar.
// Use with RawDo for providers that need custom request construction.
func (h *HttpClient) GetClient() *http.Client {
	return &http.Client{
		Transport: h.transport,
		Jar:       h.jar,
		Timeout:   h.timeout,
	}
}

// RawDo executes a caller-built request through HttpClient's UA/proxy/retry machinery.
// Used by providers that need custom request construction (e.g. yopmail scraping flow).
// bodyBytes should be provided for POST requests so retry can rewind the body.
func (h *HttpClient) RawDo(req *http.Request, bodyBytes []byte) ([]byte, error) {
	return h.do(req, bodyBytes)
}

// do executes req with UA rotation, proxy rotation, and retry on 429/network errors.
func (h *HttpClient) do(req *http.Request, bodyBytes []byte) ([]byte, error) {
	maxAttempts := len(defaultBackoff) + 1

	for attempt := 0; attempt < maxAttempts; attempt++ {
		if err := h.prepareRequest(req, bodyBytes); err != nil {
			return nil, err
		}

		client := h.getClient()
		resp, err := client.Do(req)
		if err != nil {
			if attempt < maxAttempts-1 {
				time.Sleep(defaultBackoff[attempt])
			}
			continue
		}

		body, readErr := io.ReadAll(resp.Body)
		resp.Body.Close()

		if resp.StatusCode == http.StatusTooManyRequests && attempt < maxAttempts-1 {
			time.Sleep(defaultBackoff[attempt])
			continue
		}

		if readErr != nil {
			return nil, &TempMailError{Provider: "httpclient", Message: readErr.Error()}
		}

		return body, nil
	}

	return nil, &TempMailError{Provider: "httpclient", Message: "max retries exceeded"}
}

// prepareRequest sets UA and proxy for this attempt.
func (h *HttpClient) prepareRequest(req *http.Request, bodyBytes []byte) error {
	if h.randomUA {
		h.mu.Lock()
		ua := uaPool[h.rng.Intn(len(uaPool))]
		h.mu.Unlock()
		req.Header.Set("User-Agent", ua)
	}

	proxyCount := len(h.proxies)
	if proxyCount > 0 {
		h.mu.Lock()
		proxyURL := h.proxies[h.rng.Intn(proxyCount)]
		h.mu.Unlock()

		proxy, err := url.Parse(proxyURL)
		if err != nil {
			return &TempMailError{Provider: "httpclient", Message: "invalid proxy: " + proxyURL}
		}
		h.transport.Proxy = http.ProxyURL(proxy)
	} else {
		h.transport.Proxy = nil
	}

	// Reset body for retries.
	if bodyBytes != nil {
		req.Body = io.NopCloser(bytes.NewReader(bodyBytes))
		req.ContentLength = int64(len(bodyBytes))
	}

	return nil
}

// getClient returns an http.Client using the shared transport and jar.
func (h *HttpClient) getClient() *http.Client {
	return &http.Client{
		Transport: h.transport,
		Jar:       h.jar,
		Timeout:   h.timeout,
	}
}
